### ISSUE TITLE
Support WASM Architecture

### ISSUE THREAD

**Issue by magik6k** [ipfs/kubo#5694]
So Go now Supports targeting WASM, but with certain limitations (like not having 'normal' syscalls). It's very low priority and probably a bit of work, but would be a nice 'benchmark' of portability.

First step would be to make the thing compile at all, currently stack of errors I get is down to:
```
# gx/ipfs/QmXTpwq2AkzQsPjKqFQDNY2bMdsAT53hUBETeyj8QRHTZU/sha256-simd
../../../gx/ipfs/QmXTpwq2AkzQsPjKqFQDNY2bMdsAT53hUBETeyj8QRHTZU/sha256-simd/cpuid.go:23:14: undefined: haveArmSha
../../../gx/ipfs/QmXTpwq2AkzQsPjKqFQDNY2bMdsAT53hUBETeyj8QRHTZU/sha256-simd/cpuid.go:27:16: undefined: cpuid
../../../gx/ipfs/QmXTpwq2AkzQsPjKqFQDNY2bMdsAT53hUBETeyj8QRHTZU/sha256-simd/cpuid.go:32:13: undefined: xgetbv
../../../gx/ipfs/QmXTpwq2AkzQsPjKqFQDNY2bMdsAT53hUBETeyj8QRHTZU/sha256-simd/cpuid.go:40:18: undefined: cpuid
../../../gx/ipfs/QmXTpwq2AkzQsPjKqFQDNY2bMdsAT53hUBETeyj8QRHTZU/sha256-simd/cpuid.go:44:19: undefined: cpuidex
../../../gx/ipfs/QmXTpwq2AkzQsPjKqFQDNY2bMdsAT53hUBETeyj8QRHTZU/sha256-simd/cpuid.go:52:18: undefined: cpuid
../../../gx/ipfs/QmXTpwq2AkzQsPjKqFQDNY2bMdsAT53hUBETeyj8QRHTZU/sha256-simd/cpuid.go:56:17: undefined: cpuid
../../../gx/ipfs/QmXTpwq2AkzQsPjKqFQDNY2bMdsAT53hUBETeyj8QRHTZU/sha256-simd/cpuid.go:61:14: undefined: xgetbv
../../../gx/ipfs/QmXTpwq2AkzQsPjKqFQDNY2bMdsAT53hUBETeyj8QRHTZU/sha256-simd/cpuid.go:62:20: undefined: cpuidex
../../../gx/ipfs/QmXTpwq2AkzQsPjKqFQDNY2bMdsAT53hUBETeyj8QRHTZU/sha256-simd/cpuid.go:90:16: undefined: cpuid
../../../gx/ipfs/QmXTpwq2AkzQsPjKqFQDNY2bMdsAT53hUBETeyj8QRHTZU/sha256-simd/cpuid.go:90:16: too many errors
# gx/ipfs/QmPtj12fdwuAqj9sBSTNUxBNu8kCGNp8b3o8yUzMm5GHpq/pb
../../../gx/ipfs/QmPtj12fdwuAqj9sBSTNUxBNu8kCGNp8b3o8yUzMm5GHpq/pb/pb.go:421:9: undefined: terminalWidth
../../../gx/ipfs/QmPtj12fdwuAqj9sBSTNUxBNu8kCGNp8b3o8yUzMm5GHpq/pb/pb.go:430:18: undefined: terminalWidth
# gx/ipfs/QmSvfeW68LC13nVt3BiwHKFneSa4DCdq3erG8RNtJvq7Ni/go-reuseport
../../../gx/ipfs/QmSvfeW68LC13nVt3BiwHKFneSa4DCdq3erG8RNtJvq7Ni/go-reuseport/interface.go:33:9: undefined: available
../../../gx/ipfs/QmSvfeW68LC13nVt3BiwHKFneSa4DCdq3erG8RNtJvq7Ni/go-reuseport/interface.go:50:6: undefined: available
../../../gx/ipfs/QmSvfeW68LC13nVt3BiwHKFneSa4DCdq3erG8RNtJvq7Ni/go-reuseport/interface.go:54:9: undefined: listenStream
../../../gx/ipfs/QmSvfeW68LC13nVt3BiwHKFneSa4DCdq3erG8RNtJvq7Ni/go-reuseport/interface.go:61:6: undefined: available
../../../gx/ipfs/QmSvfeW68LC13nVt3BiwHKFneSa4DCdq3erG8RNtJvq7Ni/go-reuseport/interface.go:65:9: undefined: listenPacket
../../../gx/ipfs/QmSvfeW68LC13nVt3BiwHKFneSa4DCdq3erG8RNtJvq7Ni/go-reuseport/interface.go:72:6: undefined: available
../../../gx/ipfs/QmSvfeW68LC13nVt3BiwHKFneSa4DCdq3erG8RNtJvq7Ni/go-reuseport/interface.go:102:6: undefined: available
../../../gx/ipfs/QmSvfeW68LC13nVt3BiwHKFneSa4DCdq3erG8RNtJvq7Ni/go-reuseport/interface.go:106:9: undefined: dial
```

Built with `GOOS=js GOARCH=wasm go build -tags="nofuse purego" -ldflags="-X "github.com/ipfs/go-ipfs".CurrentCommit=47e9466ac" -o "cmd/ipfs/ipfs" "github.com/ipfs/go-ipfs/cmd/ipfs"`

* sha256-simd - fixed in https://github.com/minio/sha256-simd/pull/33, so an update will fix those
* go-reuseport - we should find a way to just disable it, probably by either just plugin-ifying transports or sprinkling build tags injecting dummy things in place of the implementation. Probably a good excuse to fix some technical debt.
* [pb](https://github.com/cheggaaa/pb) (progress bar) - no idea, probably just PR a fix/workaround

Next step would be to write some transport to talk to outside nodes. `/ws`/`/wss` should be easy to wrap.

**Comment by terorie** [comment]
https://github.com/cheggaaa/pb/pull/140 should fix `pb` wasm support

**Comment by lidel** [comment]
With ipfs/boxo#935 merged, `boxo/files` is no longer a wasm blocker. The kubo binary still does not compile for wasm, but the remaining work for `GOOS=wasip1 GOARCH=wasm` is tractable (flatfs works there; wasi has a real filesystem). 

I think the times changed, and the `GOOS=js` browser case stays out of scope; see the note at the end. Happy to change my mind if someone provides a functional PoC :)

## How we could make `GOOS=wasip1 GOARCH=wasm` work

POC against current `master` (v0.41)  plus ipfs/boxo#935: `cmd/ipfs` compiles to a 106 MB `.wasm` after four changes:

1. **Remove `pebble.FormatNewest` from `config/init.go:10`.** It is the only pebble symbol used there, and the import drags all of pebble into every package that transitively imports `config`. Inline the int (`24`) or move `pebbleSpec()` into the pebble plugin.
2. **Gate the three storage plugins.** Add `//go:build !wasm` to `plugin/plugins/{pebbleds,badgerds,levelds}`, and have `plugin/loader/preload.sh` emit a wasm-only `preload.go` listing flatfs only.
3. **Gate `cmd/ipfswatch`** (it imports the storage plugins directly): change `//go:build !plan9` to `//go:build !plan9 && !wasm`.
4. **Cover `wasip1` in `cheggaaa/pb`'s `terminalWidth` fallback.** v1.0.29 has it for `appengine js` but not `wasip1`. Either upstream the one-line tag fix or replace pb in the four CLI commands that use it.

After those, `GOOS=wasip1 GOARCH=wasm go build -tags=nofuse,openssl ./cmd/ipfs` succeeds. Runtime behavior of libp2p transports under wasi is a separate question; none of it is a compile blocker. I did not open PR because it requires extra work (refactoring datastore plugin code and init so incompatible deps can be compiled out), and there are better options in 2026, like [Boxo](https://github.com/ipfs/boxo) and [Helia](https://github.com/ipfs/helia) (see below)

## Why `GOOS=js GOARCH=wasm` is out of scope, I think?

A lot of changed since 2018, in both kubo and go-libp2p. The PoC from https://github.com/ipfs/kubo/pull/5878 is no longer enough.

With new go-libp2p code surface, problems cannot be fixed inside kubo. `github.com/libp2p/go-libp2p/defaults.go` and `.../config/config.go` both import `p2p/transport/webrtc` unconditionally, and that transport does not compile under `js`: pion's `_js.go` files expose only the browser client API and miss the server-side surface go-libp2p uses (`SCTPTransport.OnClose`, `SettingEngine.SetICEUDPMux`, and friends). Closing that gap requires upstream build-tag work in go-libp2p. 

Even past compile, the browser runtime has no filesystem (flatfs is out) and no raw TCP or UDP (so `tcp`, `quic`, and `webtransport` listen-side are dead), leaving only websocket-as-client. 

That niche is exactly what [Helia](https://github.com/ipfs/helia) already covers, in a JS-native runtime an order of magnitude smaller than 100+ MB of Go-via-wasm.

## TLDR

So I think guidance for 2026Q2 is:

- if you want `GOOS=js GOARCH=wasm`, use either [Boxo](https://github.com/ipfs/boxo) directly (compiles since ipfs/boxo#935) or even better, use browser-first [Helia](https://github.com/ipfs/helia) (native JS implementation)
- if you want `GOOS=wasip1 GOARCH=wasm`, comment in this issue why (why not [Boxo](https://github.com/ipfs/boxo)  or  [Helia](https://github.com/ipfs/helia)?)

Without clear user need, WASM will remain low priority.
